src/storage/archive_index.rs

```diff
@@ -27,63 +29,87 @@ impl FileInfo {
+impl Index {
+    pub(crate) fn write_sqlite<P: AsRef<Path>>(&self, destination: P) -> Result<()> {
+        let destination = destination.as_ref();
+        if destination.exists() {
+            fs::remove_file(destination)?;
+        }
 
-    let mut archive = zip::ZipArchive::new(zipfile)?;
+        let conn = rusqlite::Connection::open(destination)?;
+        conn.execute("PRAGMA synchronous = FULL", ())?;
+        conn.execute("BEGIN", ())?;
+        conn.execute(
+            "
+            CREATE TABLE files (
+                id INTEGER PRIMARY KEY,
+                path TEXT UNIQUE,
+                start INTEGER,
+                end INTEGER,
+                compression INTEGER
+            );
+            ",
+            (),
+        )?;
 
-    let conn = rusqlite::Connection::open(&destination)?;
-    conn.execute("PRAGMA synchronous = FULL", ())?;
-    conn.execute("BEGIN", ())?;
-    conn.execute(
-        "
-        CREATE TABLE files (
-            id INTEGER PRIMARY KEY,
-            path TEXT UNIQUE,
-            start INTEGER,
-            end INTEGER,
-            compression INTEGER
-        );
-        ",
-        (),
-    )?;
+        for (name, info) in self.files.iter() {
+            conn.execute(
+                "INSERT INTO files (path, start, end, compression) VALUES (?, ?, ?, ?)",
+                (
+                    name,
+                    info.range.start(),
+                    info.range.end(),
+                    info.compression as i32,
+                ),
+            )?;
+        }
 
-    for i in 0..archive.len() {
-        let zf = archive.by_index(i)?;
+        conn.execute("CREATE INDEX idx_files_path ON files (path);", ())?;
+        conn.execute("END", ())?;
+        conn.execute("VACUUM", ())?;
+        Ok(())
+    }
 
-        let compression_bzip = CompressionAlgorithm::Bzip2 as i32;
+    pub(crate) fn from_zip<R: io::Read + io::Seek>(zipfile: &mut R) -> Result<Self> {
+        let mut archive = zip::ZipArchive::new(zipfile)?;
 
-        conn.execute(
-            "INSERT INTO files (path, start, end, compression) VALUES (?, ?, ?, ?)",
-            (
-                zf.name(),
-                zf.data_start(),
-                zf.data_start() + zf.compressed_size() - 1,
-                match zf.compression() {
-                    zip::CompressionMethod::Bzip2 => compression_bzip,
-                    c => bail!("unsupported compression algorithm {} in zip-file", c),
+        let mut index = Index {
+            files: HashMap::with_capacity(archive.len()),
+        };
+
+        for i in 0..archive.len() {
+            let zf = archive.by_index(i)?;
+
+            index.files.insert(
+                zf.name().to_owned(),
+                FileInfo {
+                    range: FileRange::new(
+                        zf.data_start(),
+                        zf.data_start() + zf.compressed_size() - 1,
+                    ),
+                    compression: CompressionAlgorithm::Bzip2,
```

Comment:
We don't need to check the compression type from the zip file?

Reply:
typically we don't, we only use one algorithm. 

But since I had the check in the previous version too, I added it back.